### PR TITLE
feat: add support for overriding subgraph routing urls

### PR DESCRIPTION
**Pull request by cs-clarence**
## Motivation and Context
Allows overriding routing urls for subgraphs via `config.yaml`:
```yaml
override_routing_url:
  subgraphs:
    some-subgraph: http://new-subgraph-routing-url.com/graphql
```

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

**Comment by StarpTech** [comment]
Hi @ cs-clarence, can you give more context on why this feature is valuable to add to the core?

The routing URL is stored in the route config which is fetched from the controlplane. The integrity of the URL is important to enrich requests with proper trace attributes. If we allow arbitrary subgraph URLs we lose the connection.
In order to change the subgraph URL you have to update the subgraph with the wgc cli.

**Comment by cs-clarence** [comment]
Hey @StarpTech, I've seen this feature in the apollo router [here](https://www.apollographql.com/docs/router/configuration/overview#subgraph-routing-urls). 

It would be useful for routers inside a k8s cluster so they can use the cluster-local DNS name while keeping the public DNS name config in the control plane. I understand that this is a niche use case, so feel free to close this if you deem this a not-useful feature.

**Comment by StarpTech** [comment]
> It would be useful for routers inside a k8s cluster so they can use the cluster-local DNS name while keeping the public DNS name config in the control plane. I understand that this is a niche use case, so feel free to close this if you deem this a not-useful feature.

This is a good use case! I'll come back later.

**Comment by jgautheron** [comment]
Got a similar use case here, currently replacing it before deployment with `sed` but would be cleaner to allow an override either by config and/or env var.

**Comment by StarpTech** [comment]
@jgautheron thanks for the feedback. Will look into it.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/internal/config/config.go [hunk 1]
```diff
@@ -83,21 +83,21 @@ type TrafficShapingRules struct {
...
 
 type BackoffJitterRetry struct {
-	Enabled     bool          `yaml:"enabled" default:"true" envconfig:"RETRY_ENABLED"`
-	Algorithm   string        `yaml:"algorithm" default:"backoff_jitter" validate:"oneof=backoff_jitter"`
-	MaxAttempts int           `yaml:"max_attempts" default:"5" validate:"required,min=1,required_if=Algorithm backoff_jitter"`
-	MaxDuration time.Duration `yaml:"max_duration" default:"10s" validate:"required,min=1s,required_if=Algorithm backoff_jitter"`
-	Interval    time.Duration `yaml:"interval" default:"3s" validate:"required,min=100ms,required_if=Algorithm backoff_jitter"`
+	Enabled     bool          `yaml:"enabled"      default:"true"           envconfig:"RETRY_ENABLED"`
+	Algorithm   string        `yaml:"algorithm"    default:"backoff_jitter"                           validate:"oneof=backoff_jitter"`
+	MaxAttempts int           `yaml:"max_attempts" default:"5"                                        validate:"required,min=1,required_if=Algorithm backoff_jitter"`
```
[the remark this answers is not in the document]
Comment (cs-clarence) [reply router/internal/config/config.go:98]: Oh I didn't realize I had it on. Will revert it

File: router/core/router.go [hunk 2]
```diff
@@ -224,9 +227,32 @@ func NewRouter(opts ...Option) (*Router, error) {
 // updateServer starts a new Server. It swaps the active Server with a new Server instance when the config has changed.
 // This method is safe for concurrent use. When the router can't be swapped due to an error the old server kept running.
 func (r *Router) updateServer(ctx context.Context, cfg *nodev1.RouterConfig) error {
+	overrides := make(map[string]string)
+
+	for _, sg := range cfg.Subgraphs {
```
Comment (StarpTech) [inline router/core/router.go:232]: Put this logic in a method for better readability e.g. `configureSubgraphOverwrites`.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/router.go [hunk 3]
```diff
@@ -588,9 +619,17 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
 
 		subgraphs := make([]Subgraph, len(routerConfig.Subgraphs))
 		for _, s := range routerConfig.Subgraphs {
-			parsedURL, err := url.Parse(s.RoutingUrl)
+			rawURL := s.RoutingUrl
```
Comment (StarpTech) [inline router/core/router.go:622]: Move the validation logic into the proposed `configureSubgraphOverwrites` and return the error when a subgraph has an invalid url. Although the URLs are already validated through the validate config package. I'm also fine with both.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/router.go [hunk 4]
```diff
@@ -473,6 +528,11 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
 	)
 
 	httpRouter := chi.NewRouter()
+	httpRouter.Use(func(h http.Handler) http.Handler {
```
Comment (StarpTech) [inline router/core/router.go:531]: Mistake? This is an empty middleware.

File: router/core/router.go [hunk 5]
```diff
@@ -221,12 +223,63 @@ func NewRouter(opts ...Option) (*Router, error) {
 	return r, nil
 }
 
+func (r *Router) configureSubgraphOverwrites(cfg *nodev1.RouterConfig) ([]Subgraph, error) {
+	overrides := make(map[string]string)
+
+	// Validate all the URLs first before we start overriding
```
Comment (StarpTech) [inline router/core/router.go:229]: Much better :+1:

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/router.go (hunk 4 above)
Comment (cs-clarence) [reply router/core/router.go:531]: Oh yeah, I forgot to remove it

**Comment by StarpTech** [comment]
@cs-clarence ready for review again?

**Comment by cs-clarence** [comment]
@StarpTech yes

**Comment by StarpTech** [comment]
I continue your work here #152 I found a couple of issues and fixed them there. I hope it's okay. I could not push the branch here.

**Comment by cs-clarence** [comment]
@StarpTech yeah that's alright, it's better for you to take over because although I tested the code, I haven't verified that it works without issues.

**Comment by StarpTech** [comment]
@cs-clarence thank you again for pushing this initiative.

**Pull request closed without merging**